developer/introduction.rst

```diff
@@ -0,0 +1,183 @@
+of user-defined classes.
+
+`runserver_plus
+<django-extensions.readthedocs.io/en/latest/runserver_plus.html>`_:
+Typical runserver with Werkzeug debugger baked in.
+
+Django Debug Toolbar
+--------------------
+
+The Django Debug Toolbar is a configurable set of panels that display
+various debug information about the current request/response and when
+clicked, display more details about the panel’s content.
+
+It can be installed with:
+
+``pip install django-debug-toolbar``
+
+More details can be found on its `website
+<https://django-debug-toolbar.readthedocs.io/en/latest/>`_
+or on `pypi <https://pypi.org/project/django-debug-toolbar/>`_.
+
+IPython and ipdb
+----------------
+
+IPython (Interactive Python) is a command shell for interactive computing
+in multiple programming languages, originally developed for the Python
+programming language, that offers introspection, rich media, shell
+syntax, tab completion, and history.
+
+It provides:
+- A powerful interactive shell.
+- A browser-based notebook interface with support for code, text,
+mathematical expressions, inline plots and other media.
+- Support for interactive data visualization and use of GUI toolkits.
+- Flexible, embeddable interpreters to load into one's own projects.
+- Tools for parallel computing.
+
+More details including installation and updates can be found on the
+`official webite <https://ipython.org>`_ and a simple usage tutorial
+can be found `here <http://cs231n.github.io/ipython-tutorial/>`_.
+
+As for ipdb, it exports functions to access the IPython debugger. It
+comes along with ipython. More details about ipdb can be obtained
+on `its pypi page <https://pypi.org/project/ipdb/>`_.
+
+OpenWISP Implementation
+-----------------------
+
+These tools can be implemented in OpenWISP modules and can significantly
+improve the workflow. Let's do a walkthrough of how to use them in
+`OpenWISP Controller <https://github.com/openwisp/openwisp-controller>`_
+as an example.
+
+For instance, instead of ``python manage.py runserver``, the command
+``python manage.py runserver_plus`` can be used which will provide more
+debugging information.
+
+Also, ``ipython`` can be used along ``shell_plus`` by running the command
+``./manage.py shell_plus --ipython`` in the terminal. This will provide
+an interactive shell for running code in python.
+
+To debug the code, ``ipdb`` can be used. Commands similar to
+``bin/ipdb3 mymodule.py`` may be used to carry out that process.
+
+Lastly, ``django-debug-toolbar`` may be used along to display information
+about processes occuring on the website. To achieve that, some code needs
+to be added to our current module, i.e ``openwisp-controller``.
+The file ``local_settings_example.py`` can be renamed to
+```local_settings.py`` and some code can be added to it so that
+``django-debug-toolbar`` works. The lines ``'debug_toolbar',`` and
+``'debug_toolbar.middleware.DebugToolbarMiddleware',`` need to be added
```

Comment:
we need a full example, test this first in your env and when you're sure it works I'll follow your instructions and test it to make sure it works